File: src/content/docs/topics/cs/coding-problems/trees/124-binary-tree-maximum-path-sum.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func maxPathSum(root *TreeNode) int {
	// TODO: implement
	return 0
}

func runTests() {
	// [1,2,3] -> 6
	assert(maxPathSum(&TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}) == 6)

	// [-10,9,20,null,null,15,7] -> 42
	assert(maxPathSum(&TreeNode{Val: -10,
		Left:  &TreeNode{Val: 9},
		Right: &TreeNode{Val: 20, Left: &TreeNode{Val: 15}, Right: &TreeNode{Val: 7}},
	}) == 42)

	// single node
	assert(maxPathSum(&TreeNode{Val: 1}) == 1)

	// all negative: pick least negative
	assert(maxPathSum(&TreeNode{Val: -3, Left: &TreeNode{Val: -1}, Right: &TreeNode{Val: -2}}) == -1)

	// negative root, positive children
	assert(maxPathSum(&TreeNode{Val: -1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}) == 4)

	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/trees/124-binary-tree-maximum-path-sum.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val: number = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
        this.val = val; this.left = left; this.right = right;
    }
}

function buildTree(vals: (number | null)[]): TreeNode | null {
    if (!vals.length) return null;
    const root = new TreeNode(vals[0] as number);
    const q: TreeNode[] = [root];
    let i = 1;
    while (q.length && i < vals.length) {
        const node = q.shift()!;
        if (i < vals.length && vals[i] !== null) {
            node.left = new TreeNode(vals[i] as number);
            q.push(node.left);
        }
        i++;
        if (i < vals.length && vals[i] !== null) {
            node.right = new TreeNode(vals[i] as number);
            q.push(node.right);
        }
        i++;
    }
    return root;
}

function maxPathSum(root: TreeNode | null): number {
    // TODO: implement
    return 0;
}

function _runTests(): void {
    assert(maxPathSum(buildTree([1, 2, 3])) === 6);
    assert(maxPathSum(buildTree([-10, 9, 20, null, null, 15, 7])) === 42);
    assert(maxPathSum(buildTree([1])) === 1);
    assert(maxPathSum(buildTree([-3, -1, -2])) === -1);
    assert(maxPathSum(buildTree([-1, 2, 3])) === 4);
    // perf
    function makeTree(n: number): TreeNode | null {
        if (!n) return null;
        const nodes = Array.from({ length: n }, (_, i) => new TreeNode(i));
        for (let i = 0; i < n; i++) {
            if (2 * i + 1 < n) nodes[i].left = nodes[2 * i + 1];
            if (2 * i + 2 < n) nodes[i].right = nodes[2 * i + 2];
        }
        return nodes[0];
    }
    const root = makeTree(131_071);
    const t0 = performance.now();
    maxPathSum(root);
    console.log(`perf max_path_sum on 131071-node complete tree: ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();
